---
title: "BMEG 400E: Assignment 1"
output:
  html_document:
    toc: true
    toc_depth: 4
    toc_float: true
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Authors and contributions

Following completion of your assignment, please fill out this section with the authors and their contributions to the assignment.  If you worked alone, only the author (e.g. your name and student ID) should be included.

Authors: Name1 (studentID1) and Name2 (studentID2)

Contributions: (example) N1 and N2 worked together on the same computer to complete the assignment. N1 typed for the first half and N2 typed for the second half. 

Both Jason Fung and Ariel Huynh worked on the same computer. Ariel primarily did questions 1 and 2 by getting familiar with the server and setting up conda environments while Jason overlooked and searched up unix commands. Jason continued the assignment from question 3 onwards while Ariel reviewed the results and unix commands. 


# Assignment Overview

For this assignment we will be using data from a Chromatin Immunoprecipitation sequencing (ChIP-seq) produced by Choi, Won-Young, et al., (2020). *https://pubmed.ncbi.nlm.nih.gov/33293480/*. This data is part of a series of experiments that aimed to understand the chromatin changes that happen when an induced Pluripotent Stem Cell (iPSC) undergo differentiation towards a Neuronal Progenitor Cell (NPC). The **fastqc** files from this experiment have been pre-downloaded and added to the server, under the following path: **/projects/bmeg/A1/**. The datasets that we will be using for the assignment are in a shared *reading only* location that you can read but ARE NOT allowed to alter. 

This assignment has 3 main goals: 

  1. Get familiar with the server
  
  2. Manage a conda environment
  
  3. Perform an alignment against the human genome
  
  4. Cleaning your folder and uploading your assignment to Github


IMPORTANT: Every time you see *#?#* means, that's a question you need to answer, with a command or with an interpretation of the data. When an instruction follows #?# you need to type the command you used to complete the task. Please make sure to check for them and answer them all, as this is how your assignment will be graded. 


## 1. Getting Familiar with the Server

For our course we will be using a server based on a Docker/Linux system. This will be the place where you are gonna do the on-class practices and the assignments. There are a couple of things that you need to be aware while working on the server. The server has limited storage and computer power. Therefore, you will be required to be mindful of the processes you run as well as the files you keep. 

To join the server you will need to be on an active connection of a UBC Virtual Private Network (VPN). If you do not have one already, you can check how to install it here: *https://it.ubc.ca/services/email-voice-internet/myvpn/setup-documents*. Once the VPN has been set, you will need to open a terminal. 

  - **Windows system:** 
      a. Install a terminal emulator like Putty (*https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html*). 
      b. This will allow you to do a SSH connection using the server's IP (orca1.bcgsc.ca) and your credentials. 
  
  - **Linux/Unix system (Apple computer and Ubuntu):** 
      a. Open a terminal
      b. Type the following command: ssh username@orca1.bcgsc.ca 
      c. When prompted, type your password to complete the login 
      
    
Now that you have successfully joined the server, let's get you used to some basic commands

### a. Check a directory's path 

When using terminal, paths work as the addresses of the files and directories you want to access and easily move between them. Once you have logged into the server you will be in your *home directory*.

```{bash, eval=FALSE}
#?# Check your current directory's path: 0.5 pt
Use command: pwd
/home/jfung_bmeg22

```


### b. Creating a directory 

```{bash, eval=FALSE}

#?# On your home directory create a new directory (folder) - 1pt
## This folder will host the files created throughout this assignment 
mkdir assignment-1
current directory: /home/jfung_bmeg22/assignment-1

```

*Note:* Generally speaking, is good to follow naming convention when using the terminal. Remember: 

  - Do not start a name with a number

  - Names are case sensitive (ASSIGNMENT1.txt and assignment1.txt are not the same)

  - Avoid to use spaces, as they are interpreted as special characters by the computer

  - Use _ or - to replace spaces

  - File extensions are meaningful for us to know the file format but in terminal you can use it as part of the file name. Ex: you will be able to open a .sequences file that has tab delimited information
  

### c. Moving within directories

Access your newly created directory

```{bash, eval=FALSE}
#?# Type the command you need to use to move to your newly created directory - 0.5 pt 
## Tip: look at the *cd* command documentation

cd assignment-1

```

How would you move back to your home directory (*/home/your_user_name*)?

```{bash, eval=FALSE}

#Check the tutorial: https://www.computerhope.com/unix/ucd.htm 

#?# Using the complete directory path of your home directory: - 0.25 pt
use command cd /home/jfung_bmeg22/ to go back to home directory

#?# Using the "go to parent directory" shortcut - 0.25
use command cd ~


```

### d. Explore the dataset 

The sequencing data that we will be using is paired-end. This means that each sequence has been sequenced twice, one on each end (5' and 3'). Choose **one** of the reads files (1 or 2) for the following exercises.


```{bash, eval=FALSE}

#?# Look at the first 3 rows of the dataset - 0.25
# Tip: look at the *head* command documentation 
Using the head command, we get the following results on the command line:
@SRR12694366.1.1000000
CCACAGCCATGCCTCCAAGGGTCCTTGATCCAAATAGCAGAACAAATCACATAAACTCTCAGGGCCTACTTTAGAAAGGCATGTGACTTGAGTTTCTGTATCGACTGTTCCTCTTGTCCAGAGGCCTTAATTAATCCAGGAATGGCGGCAT
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,
@SRR12694366.1.10000029
TTGGAAAGTCTTGCCAGTCTGACGACATTAGCCACAGTAGACTATTGTCAAGATGTGGGCAGACTATGTGTTTGCTAATCATGTCTCTTTACTCCGTCATAATTCCTCAGGGCACAAGATTAATAGATCGGAAGAGCACACGTCTGAACTC
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF::FFFFFF:FF,:FFFFFFFFFFFFFFFFF
@SRR12694366.1.10000035
GAGCCAGATTCTGCACAAGGGTTGGGTGCCACAGGCTGGCCCAGTGGCATGGCCTGGTAAGCTCCTTAGCCAGGACAGCCAGGCAGGTTGTCATCATCTGCTGTTCCTCCCTCAGCCCTCAGTGTGTATCTGCAGCAGGGACTTCTGTGCA



#?# Look at the last 7 rows of the dataset - 0.25
# Tip: look at the *tail* command documentation 
+
Similarly to the head command, 
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFF:F:FFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFF:F:FFF:F:FFFF:FFFFFFF,FFFFFFFFFFFFFF
@SRR12694366.1.9999995
TTATAGCATTATCATATATCTGTGCCAATAAGGGCTATAGATTTTTGTTTTAAATAAAAACAAGGGGCCAGGCATAGTAATTCATGCCTGTAATCCCTGTGTTTTGGGAGGCTGAGGCAAGAGAATCACTTGGCCAGGAGTTCAAGACCAG
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF
@SRR12694366.1.9999998
CCGAAGATCCTTACCATCTGTGCAACCAAGTGGGTGGGACAGTGGAGGTGGGAGCCTCCACTATTTTGAACCTCAAGTTTCTTGGAAAATTTCTGTCAGACTTAGATGAGATCATTCATGTGAAGTTGATTTGTGAAAGGTCTAGAATGAG
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
#?# Explore the file with the *less* command  - 0.25
Using the less command: less input_iPSC_SRA66_subset_1.fastqc, the output allows the user to scroll through the fastq file. Therefore, the user views similar output structure from the head and tail commands, but are not limited to only the head and tail rows of the file.

```

### e. Piping 

Because this is a very large dataset we will proceed to subset it in order to make it more manageable for the assignment. Using the commands that you learned above:

```{bash, eval=FALSE}

#?# How many lines does the file was in total? - 0.25
#Tip: look at the *wc* command documentation
Typing the command: wc -l input_iPSC_SRA66_subset_1.fastq, there is a total of 10459640 lines in the file

#?# Select only the id lines (e.g. @SRR12506919.667552 667552 length=151) of the dataset (the ones that start with @ and are followed by the read id) - 0.75 pt
## Tip: look at the *grep* command
The following command: grep @SRR12694366 input_iPSC_SRA66_subset_1.fastq returns only the id lines of the dataset

#?# How many reads are in the file (i.e., how many id lines are in the file)? - 1 pt
## Tip: Try using * | head * after the command line you use for the previous question
Typing the command: grep @SRR12694366 input_iPSC_SRA66_subset_1.fastq | wc -l, there is a total of 2614910 id lines in the file

#?# Select only the reads id (e.g. @SRR12506919.667552) from the id lines - 1 pt
## Tip: Look into the *cut* command. Carefully read the default delimiter, is it the case for our file?
Typing the command: grep @SRR12694366 input_iPSC_SRA66_subset_1.fastq | cut -d '.' -f1,3 gets rid of the '.1.' from the read ids

```

### f. Saving an output 


```{bash, eval=FALSE}

#?# Save a file that contains only the reads ids (the result of our previous exercise). - 0.5 pt
using command: grep @SRR12694366 /projects/bmeg/A1/input_iPSC_SRA66_subset_1.fastq | cut -d '.' -f1,3 >> read_ids

#?# Now, list all the files in a directory: 
using command: ls or ls -l
read_ids

#?# What do you see? Was the subset file created correctly? - 0.25 pt
## ----
There exists only one file in my own directory which is read_ids. I can check that the file was created correctly by using the commands head, tail, or less to check the contents inside the file.

```

### g. Creating a backup 

There will be times where you will want to save a copy of a dataset or file before modifying in case something goes wrong. For those cases, you can create a copy of any file or directory using the "copy" command

```{bash, eval=FALSE}

#?# Create a copy of the reads ids file - 0.25 pt
## Tip: man cp 
make new directory: mkdir backup 
using the command: cp read_ids read_ids_copy
#?# Change the name of the backup reads id file - 0.25 pt
## Tip: man mv
using mv read_ids_copy read_ids_backup renames the back up reads 
#?# Delete the copy of the reads id file - 0.25 pt
No point in removing the copy of the reads file when there is no copy that exists after renaming the file
However, the command is: rm read_ids_copy if there is a copy file
```

## 2. Managing Conda Environments


### a. Installing conda 

Conda is a package and environment manager (*https://docs.conda.io/en/latest/*). It helps to create different virtual environments where you can have several packages installed to meet different needs. 

```{bash, eval=FALSE}

## 1. Move to your home directory 

#?# 2. Use the wget command to get the following file: https://repo.anaconda.com/archive/Anaconda3-2021.11-Linux-x86_64.sh
wget https://repo.anaconda.com/archive/Anaconda3-2021.11-Linux-x86_64.sh

## The file extension .sh pertains to a bash script, the script you just downloaded contains all the instructions needed to install Conda to your home directory
#?# 3. Run the Anaconda3-2021.11-Linux-x86_64.sh bash script. IMPORTANT: say yes to all the steps when prompted! 
bash Anaconda3-2021.11-Linux-x86_64.sh


## 4. To finalize the installation you need to log out your current session and log in back to the server



```


### b. Create a conda environment

In order to run the reads alignments against the human genome, there are a few tools that we will need:

  - fastQC (*https://www.bioinformatics.babraham.ac.uk/projects/fastqc/*): comprehensive quality control measures of sequencing data.
  
  - bowtie2 (*http://bowtie-bio.sourceforge.net/bowtie2/index.shtml*): alignments of sequencing data. 
  
  - samtools (*http://www.htslib.org/doc/samtools.html*): set of tools to manage sam/bam alignment files 

  - htop & screen: bash system tools to visualize the server capacity and to run commands in the background, respectively. 
  
  
To install them, we will be making use of the conda environments. Conda allows you to create and manage environments for any programming language. Managing this environments mean that you can work with specific versions of different programs at any given moment, simply by loading the desired environment. You can find more information about this resource here: *https://docs.conda.io/en/latest/* . 

```{bash, eval=FALSE}
#?# Create a new conda environment: - 0.5 pt
# Tip: Consult the previously provide links or consult the conda create help (conda create --help)
conda create --name GI-env (labeled as GenomeInformatics-Env)
conda activate GI-env

```


### b. Add programs to your conda environment

Now that the environment has been created, its time to add the packages that we will need. Conda has an active community and a great documentation (*https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html*) which you can use throughout the course to help answer any questions you may have. 


```{bash, eval=FALSE}

#?# Add fastQC bowtie2 to your conda environment: - 1 pt
conda install -c bioconda fastqc
conda install -c bioconda bowtie2
## Run the following command to install samtools onto your conda environment:

#?# Add screen to your conda environment: - 1pt 
conda install -c conda-forge screen

#?# Add htop to your conda environment: - 1pt
conda install -c conda-forge htop

```

## 3. Performing Alignments 

### a. Data quality check

We will use the widely used fastQC software to do a quick inspection of the data quality. Once it has ran, it will give you an html report of the quality of the data, that you can open using a web browser. More information on how to read the output can be found here: *https://dnacore.missouri.edu/PDF/FastQC_Manual.pdf* and in the tool's website.

  
```{bash, eval=FALSE}

#?# Run fastQC on the fastq files: - 1pt

Running this command: fastqc /projects/bmeg/A1/input_iPSC_SRA66_subset_1.fastq --outdir=/home/jfung_bmeg22/assignment-1 outputs the fastqc file in html into my assignment directory

## In order to open the html files on your web browser you will need to download the files to your computer
## Ubuntu/Linux and Mac users: look at the *scp* command
## Windows users: follow the following instructions: https://stackoverflow.com/questions/6217055/how-can-i-copy-a-file-from-a-remote-server-to-using-putty-in-windows

#?# What can you say about the data quality? - 2 pt 
# ---
After transfering the fastqc files from my remote login to my personal machine using: scp jfung_bmeg22@orca1.bcgsc.ca:/home/jfung_bmeg22/assignment-1/input_iPSC_SRA66_subset_1_fastqc.html /Users/jasonfung/Downloads
, the HTML file showed that the data quality of the first read of the sequence was good. This was indicated by the high quality scores along the position in the reads from the per base sequence quality chart. 

```
  
  
### b. Running a process on the background: screen 

The processes that we are about to run, can take a long time to finish. Thus, we will make use of the *screen* tool, which allows you to run a process in the background while continue using your server session without the risk of ending the process due to a bad internet connection, accidentally closing the tab or other random circumstances. 

```{bash, eval=FALSE}

## To run a process in a background screen with screen you:

# 1. Run the following command: 
script /dev/null

# 2. Activate your conda environment 

# 3. Start a background process with a specific name 
screen -S background_screen_name 

# 4. Run the process and any commands you wish
wc -l usr/local/share/data/assignment_1/SRR12506919_1_subset.fastq

# 5. Get out of the background screen, you will need to type the following:
ctrl + A + D 

# 6. Return to the background screen to check the process
screen -r background_screen_name

# 7. Terminate the background screen once the process has ended
# Within the background screen type:
exit 

```


### c. Checking the server capacity and your current runs: htop

Another way to check on the progress of your runs and the server status is with the *htop* command. This will open a screen showing all the processes that are being currently being run in the server. Our server only has 2 cpu's/cores, the green bar next to each code, represents how much of that node it is currently in use. Always make sure to check the processes that are being run before sending yours.

```{bash, eval=FALSE}

#?# Use the htop command to describe the status of the server - 1.5 pt
Both CPU and memory are not under heavy load. Current uptime is 143 days, with a load average of 3.5-3.8. Currently there are 6 tasks/processes that are running. 

```



### d. Genome indexing - bowtie2

Now, we will need to create an index of the human genome that bowtie2 will use to map our sequences in the genome. In order to do this, you will need to use the previously downloaded files of the human genome with the desired build (e.g. hg19,hg38), you can find those files within the server here: */usr/local/share/human/*

**BEFORE RUNNING ANYTHING**: go to the "Other resources" section at the end of the assignment!


```{bash, eval=FALSE}

## Something useful to do when using a new software is to look at the documentation using the *help* option
## Try running: 
bowtie2 -h 

## IMPORTANT!!!!
## BEFORE RUNNING: go to the "Other resources" section at the end of the assignment!!
## __________________________________________________________________________________
#?# Use the hg38 build to create an index of the human genome with bowtie2
in the hg38 project folder /projects/bmeg/indexes/hg38,
use the command: bowtie2-build hg38.fa hg38_bowtie2_index

## Tip: look into the bowtie2-build help (bowtie2-build --help)  - 1.5 pt



```

### e. Alignment

We are working with paired-end data. Thus, you will need to make sure to use both fastq files to align the sequences to the genome.
**IMPORTANT:** Run with *default parameters* DO NOT specify any non-essential paramaters.

**Time flag**: This step will take up to 30 mins 

```{bash, eval=FALSE}

#?# Perform a paired-end alignment of the fastq sequences provided (located here: /projects/bmeg/A1/) to the human genome index build (located here: /projects/bmeg/indexes/hg38/ ) - 2pt
using the command: 
bowtie2 -x /projects/bmeg/indexes/hg38/hg38_bowtie2_index -1 /projects/bmeg/A1/input_iPSC_SRA66_subset_1.fastq -2 /projects/bmeg/A1/input_iPSC_SRA66_subset_2.fastq -S hg38_aligned.sam

The output is:
2614910 reads; of these:
  2614910 (100.00%) were paired; of these:
    220030 (8.41%) aligned concordantly 0 times
    2129747 (81.45%) aligned concordantly exactly 1 time
    265133 (10.14%) aligned concordantly >1 times
    ----
    220030 pairs aligned concordantly 0 times; of these:
      169387 (76.98%) aligned discordantly 1 time
    ----
    50643 pairs aligned 0 times concordantly or discordantly; of these:
      101286 mates make up the pairs; of these:
        34076 (33.64%) aligned 0 times
        30311 (29.93%) aligned exactly 1 time
        36899 (36.43%) aligned >1 times
99.35% overall alignment rate



## Tip: look at the bowtie2 --help or the documentation on the tool website (http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#getting-started-with-bowtie-2-lambda-phage-example)



```


### f. Viewing the alignments

Now, we will make use of **samtools** to review some basic features of the alignment. For mor information about this tool: *http://www.htslib.org/doc/*

```{bash, eval=FALSE}

## Use *samtools view* to get the:
## Tip: check out the different flag (-f and -F) options
## Tip: Read the samtools view --help, read carefully for an option that allows you to *count* the results of your search

#?# Number of mapped reads - 1 pt
After sorting the alignment using: samtools sort hg38_aligned.sam aligned_sorted.sam
using: samtools view -c -f 4 aligned_sorted.sam

The output is 5195744 mapped reads

#?# Mumber of unmapped reads - 1 pt
Similarly using: samtools view -c -F 4 aligned_sorted.sam

The output is 34076 unmapped reads

```


## 4. Cleaning and Uploading

### a. Cleaning your folders

Before signing up, we need to make sure that we won't leave behind any big files that can take up a lot of memory from our server. To do this, make sure to:

  a. Delete any copies of the input assignment files you might have done on your personal folder

  b. Zip or delete all the files used for the assignment 


### b. Uploading to Github

  1. Make sure to create a Github account of your own. 

  2. Once you have set it up, create a new private repository for the course. 

  3. Use the "upload file" option to upload your assignment Rmd to the website. 

  4. Upload the link of your github repository where the assignment file is located onto Canvas

## Other Resources

### a. Bowtie2 index

Indexes take a lot of computational resources and time to run. The one we need for the human genome will take around 3 hours to be done. ***DO NOT RUN THE INDEX COMMAND***. Go on to the next step, using the previously run index: *hg38_bowtie2_index* under the following path: */projects/bmeg/indexes/*


### b. Other

You can get up to 20 points, if you answer all the questions correctly. Make sure to consult the resources given through the assignment, they are meant to make the assignment easier. 